File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmUpdater.java
package nerd.tuxmobil.fahrplan.congress.alarms;

import android.app.AlarmManager;
import android.support.annotation.NonNull;

import nerd.tuxmobil.fahrplan.congress.utils.ConferenceTimeFrame;

public class AlarmUpdater {

    public interface OnAlarmUpdateListener {

        void onCancelAlarm();

        void onRescheduleAlarm(long interval, long nextFetch);

        void onRescheduleInitialAlarm(long interval, long nextFetch);

    }

    private final ConferenceTimeFrame conference;

    private final OnAlarmUpdateListener listener;

    public AlarmUpdater(@NonNull ConferenceTimeFrame conferenceTimeFrame,
                        @NonNull OnAlarmUpdateListener listener) {
        this.conference = conferenceTimeFrame;
        this.listener = listener;
    }

    public long calculateInterval(long time, boolean initial) {
        long interval;
        long nextFetch;

        long TWO_HOURS = 2 * AlarmManager.INTERVAL_HOUR;
        long ONE_DAY = AlarmManager.INTERVAL_DAY;

        if (conference.contains(time)) {
            interval = TWO_HOURS;
            nextFetch = time + interval;
        } else if (conference.endsBefore(time)) {
            listener.onCancelAlarm();
            return 0;
        } else {
            interval = ONE_DAY;
            nextFetch = time + interval;
        }

        long shiftedTime = time + ONE_DAY;
        if (conference.startsAfter(time) && conference.startsAtOrBefore(shiftedTime)) {
            interval = TWO_HOURS;
            nextFetch = conference.getFirstDayStartTime();
            if (!initial) {
                listener.onCancelAlarm();
                listener.onRescheduleAlarm(interval, nextFetch);
            }
        }

        if (initial) {
            listener.onCancelAlarm();
            listener.onRescheduleInitialAlarm(interval, nextFetch);
        }
        return interval;
    }

}

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmUpdater.kt
package nerd.tuxmobil.fahrplan.congress.alarms

import android.app.AlarmManager
import info.metadude.android.eventfahrplan.commons.logging.Logging
import info.metadude.android.eventfahrplan.commons.temporal.Moment
import nerd.tuxmobil.fahrplan.congress.repositories.AppRepository
import nerd.tuxmobil.fahrplan.congress.utils.ConferenceTimeFrame

class AlarmUpdater(

    private val conference: ConferenceTimeFrame,
    private val listener: OnAlarmUpdateListener,
    private val appRepository: AppRepository = AppRepository,
    private val logging: Logging = Logging.get()

) {

    interface OnAlarmUpdateListener {
        fun onCancelUpdateAlarm()
        fun onScheduleUpdateAlarm(interval: Long, nextFetch: Moment)
    }

    companion object {

        private const val LOG_TAG = "AlarmUpdater"
        private const val TWO_HOURS = 2 * AlarmManager.INTERVAL_HOUR
        private const val ONE_DAY = AlarmManager.INTERVAL_DAY
    }

    fun calculateInterval(moment: Moment, isInitial: Boolean): Long {
        val refreshIntervalDefaultValue = appRepository.readScheduleRefreshIntervalDefaultValue()
        val refreshInterval = appRepository.readScheduleRefreshInterval()
        val shouldUseDevelopmentInterval = refreshInterval != refreshIntervalDefaultValue
        val developmentRefreshInterval = refreshInterval.toLong()

        if (shouldUseDevelopmentInterval) {
            logging.d(LOG_TAG, "Schedule refresh interval = $developmentRefreshInterval")
            listener.onCancelUpdateAlarm()
            val nextFetch = moment.plusMilliseconds(developmentRefreshInterval)
            listener.onScheduleUpdateAlarm(developmentRefreshInterval, nextFetch)
            return developmentRefreshInterval
        }

        var interval: Long
        var nextFetch: Moment
        when {
            conference.contains(moment) -> {
                logging.d(LOG_TAG, "START <= moment < END")
                interval = TWO_HOURS
                nextFetch = moment.plusMilliseconds(interval)
            }
            conference.endsAtOrBefore(moment) -> {
                logging.d(LOG_TAG, "START < END <= moment")
                listener.onCancelUpdateAlarm()
                return 0
            }
            else -> {
                logging.d(LOG_TAG, "moment < END")
                interval = ONE_DAY
                nextFetch = moment.plusMilliseconds(interval)
            }
        }
        val shiftedTime = moment.plusDays(1)
        if (conference.startsAfter(moment) && conference.startsAtOrBefore(shiftedTime)) {
            logging.d(LOG_TAG, "moment < START && START <= shiftedTime")
            interval = TWO_HOURS
            nextFetch = conference.firstDayStartTime
            if (!isInitial) {
                listener.onCancelUpdateAlarm()
                listener.onScheduleUpdateAlarm(interval, nextFetch)
            }
        }
        if (isInitial) {
            listener.onCancelUpdateAlarm()
            listener.onScheduleUpdateAlarm(interval, nextFetch)
        }
        return interval
    }

}
